README.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> Redis components for Dart

This module is a part of the [Pip.Services](http://pipservices.org) polyglot microservices toolkit.

The module for working with a Redis database of type key-value contains components for working with cache and locks: RedisCache and RedisLock.

The module contains the following packages:

- **Build** - factory default
- **Cache** - Redis Cache Components
- **Lock** - components of working with locks in Redis

<a name="links"></a> Quick links:

* [Configuration](https://www.pipservices.org/recipies/configuration)
* [API Reference](https://pub.dev/documentation/pip_services3_redis/latest/pip_services3_redis/pip_services3_redis-library.html)
* [Change Log](CHANGELOG.md)
* [Get Help](https://www.pipservices.org/community/help)
* [Contribute](https://www.pipservices.org/community/contribute)


## Use

Add this to your package's pubspec.yaml file:
```yaml
dependencies:
  pip_services3_redis: version
```

Now you can install package from the command line:
```bash
pub get
```

## Develop

For development you shall install the following prerequisites:
* Dart SDK 2
* Visual Studio Code or another IDE of your choice
* Docker

Install dependencies:
```bash
pub get
```

Run automated tests:
```bash
pub run test
```

Generate API documentation:
```bash
./docgen.ps1
```

Before committing changes run dockerized build and test as:
```bash
./build.ps1
./test.ps1
./clear.ps1
```

## Contacts

The library is created and maintained by:
- **Sergey Seroukhov**
- **Levichev Dmitry**

The documentation is written by:
- **Mark Makarychev**
- **Levichev Dmitry**





CHANGELOG.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> Redis components for Dart Changelog

## 1.0.0

- Initial version, created by Sergey Seroukhov and Dmitry Levichev

## 1.1.0

- Migrated to null safety


test/lock/RedisLock.test.dart
// import { ConfigParams } from 'package:pip_services3_commons-node';

// import { RedisLock } from '../../src/lock/RedisLock';
// import { LockFixture } from '../fixtures/LockFixture';

// suite('RedisLock', ()=> {
//     var _lock: RedisLock;
//     var _fixture: LockFixture;

//     setup((done) => {
//         let host = process.env['REDIS_SERVICE_HOST'] || 'localhost';
//         let port = process.env['REDIS_SERVICE_PORT'] || 6379;

//         _lock = new RedisLock();

//         let config = ConfigParams.fromTuples(
//             'connection.host', host,
//             'connection.port', port
//         );
//         _lock.configure(config);

//         _fixture = new LockFixture(_lock);

//         _lock.open(null, done);
//     });

//     teardown((done) => {
//         _lock.close(null, done);
//     });

//     test('Try Acquire Lock', (done) => {
//         _fixture.testTryAcquireLock(done);
//     });

//     test('Acquire Lock', (done) => {
//         _fixture.testAcquireLock(done);
//     });

//     test('Release Lock', (done) => {
//         _fixture.testReleaseLock(done);
//     });

// });


test/lock/RedisLock_test.dart
import 'dart:io';
import 'package:test/test.dart';
import 'package:pip_services3_commons/pip_services3_commons.dart';

import 'package:pip_services3_redis/pip_services3_redis.dart';
import '../fixtures/LockFixture.dart';

void main() {
  group('RedisLock', () {
    late RedisLock _lock;
    late LockFixture _fixture;

    setUp(() async {
      var host = Platform.environment['REDIS_SERVICE_HOST'] ?? 'localhost';
      var port = Platform.environment['REDIS_SERVICE_PORT'] ?? 6379;

      _lock = RedisLock();

      var config = ConfigParams.fromTuples(
          ['connection.host', host, 'connection.port', port]);
      _lock.configure(config);

      _fixture = LockFixture(_lock);

      await _lock.open(null);
    });

    tearDown(() async {
      await _lock.close(null);
    });

    test('Try Acquire Lock', () async {
      await _fixture.testTryAcquireLock();
    });

    test('Acquire Lock', () async {
      await _fixture.testAcquireLock();
    });

    test('Release Lock', () async {
      await _fixture.testReleaseLock();
    });
  });
}


test/fixtures/LockFixture.dart
import 'package:test/test.dart';
import 'package:pip_services3_components/pip_services3_components.dart';

final LOCK1 = 'lock_1';
final LOCK2 = 'lock_2';
final LOCK3 = 'lock_3';

class LockFixture {
  final ILock _lock;

  LockFixture(ILock lock) : _lock = lock;

  Future testTryAcquireLock() async {
    // Try to acquire lock for the first time
    var result = await _lock.tryAcquireLock('123', LOCK1, 3000);
    expect(result, isTrue);

    // Try to acquire lock for the second time
    result = await _lock.tryAcquireLock('123', LOCK1, 3000);
    expect(result, isFalse);

    // Release the lock
    await _lock.releaseLock('123', LOCK1);

    // Try to acquire lock for the third time
    result = await _lock.tryAcquireLock('123', LOCK1, 3000);
    expect(result, isTrue);

    await _lock.releaseLock('123', LOCK1);
  }

  Future testAcquireLock() async {
    // Acquire lock for the first time
    await _lock.acquireLock('123', LOCK2, 3000, 1000);
    // Acquire lock for the second time
    var err;
    try {
      await _lock.acquireLock('123', LOCK2, 3000, 1000);
    } catch (ex) {
      err = ex;
    }
    expect(err, isNotNull);

    // Release the lock
    await _lock.releaseLock('123', LOCK2);

    // Acquire lock for the third time
    await _lock.acquireLock('123', LOCK2, 3000, 1000);

    await _lock.releaseLock('123', LOCK2);
  }

  Future testReleaseLock() async {
    // Acquire lock for the first time
    var result = await _lock.tryAcquireLock('123', LOCK3, 3000);
    expect(result, isTrue);

    // Release the lock for the first time
    await _lock.releaseLock('123', LOCK3);

    // Release the lock for the second time
    await _lock.releaseLock('123', LOCK3);
  }
}


lib/src/lock/RedisLock.dart
import 'dart:async';

import 'package:redis/redis.dart' as redis;
import 'package:pip_services3_commons/pip_services3_commons.dart';
import 'package:pip_services3_components/pip_services3_components.dart';

///Distributed lock that is implemented based on Redis in-memory database.
///
///### Configuration parameters ###
///
/// - [connection(s)]:
///  - [discovery_key]:         (optional) a key to retrieve the connection from [IDiscovery](https://pub.dev/documentation/pip_services3_components/latest/pip_services3_components/IDiscovery-class.html)
///  - [host]:                  host name or IP address
///  - [port]:                  port number
///  - [uri]:                   resource URI or connection string with all parameters in it
/// - [credential(s)]:
///  - [store_key]:             key to retrieve parameters from credential store
///  - [username]:              user name (currently is not used)
///  - [password]:              user password
/// - [options]:
///  - [retry_timeout]:         timeout in milliseconds to retry lock acquisition. (Default: 100)
///  - [retries]:               number of retries (default: 3)
///
///### References ###
///
/// - *:discovery:*:*:1.0        (optional) [IDiscovery](https://pub.dev/documentation/pip_services3_components/latest/pip_services3_components/IDiscovery-class.html) services to resolve connection
/// - *:credential-store:*:*:1.0 (optional) Credential stores to resolve credential [ICredentialStore](https://pub.dev/documentation/pip_services3_components/latest/pip_services3_components/ICredentialStore-class.html)
///
///### Example ###
///```dart
///    var lock = RedisLock();
///    lock.configure(ConfigParams.fromTuples([
///      'host', 'localhost',
///      'port', 6379
///    ]));
///
///    await lock.open('123');
///      ...
///
///    await lock.acquire('123', 'key1');
///
///    try {
///        // Processing...
///    } finally {
///       await lock.releaseLock('123', 'key1');
///        // Continue...
///    }
///```
class RedisLock extends Lock
    implements IConfigurable, IReferenceable, IOpenable {
  final _connectionResolver = ConnectionResolver();
  final _credentialResolver = CredentialResolver();

  final String _lock = IdGenerator.nextLong();
  int _timeout = 30000;
  int _retries = 3;

  redis.Command? _client;

  ///Configures component by passing configuration parameters.
  ///
  /// - [config]    configuration parameters to be set.
  @override
  void configure(ConfigParams config) {
    _connectionResolver.configure(config);
    _credentialResolver.configure(config);
    _timeout = config.getAsIntegerWithDefault('options.timeout', _timeout);
    _retries = config.getAsIntegerWithDefault('options.retries', _retries);
  }

  ///Sets references to dependent components.
  ///
  /// - [references] 	references to locate the component dependencies.
  @override
  void setReferences(IReferences references) {
    _connectionResolver.setReferences(references);
    _credentialResolver.setReferences(references);
  }

  ///Checks if the component is opened.
  ///
  ///Returns true if the component has been opened and false otherwise.
  @override
  bool isOpen() {
    return _client != null;
  }

  ///Opens the component.
  ///
  /// - [correlationId] 	(optional) transaction id to trace execution through call chain.
  /// Return 			Future that receives an null no errors occured.
  /// Throws error
  @override
  Future open(String? correlationId) async {
    var connection = await _connectionResolver.resolve(correlationId);
    if (connection == null) {
      throw ConfigException(
          correlationId, 'NO_CONNECTION', 'Connection is not configured');
    }
    //credential = await _credentialResolver.lookup(correlationId);

    var redisConn = redis.RedisConnection();

    //TODO: Fix work with uri connection string and credentials
    // if (connection.getUri() != null) {
    //   var url = connection.getUri();
    // } else {
    var host = connection.getHost() ?? 'localhost';
    var port = connection.getPort() ?? 6379;
    _client = await redisConn.connect(host, port);
    // }

    // if (credential != null) {
    //   var password = credential.getPassword();
    // }
  }

  ///Closes component and frees used resources.
  ///
  /// - [correlationId] 	(optional) transaction id to trace execution through call chain.
  /// Return 			Future that receives an null no errors occured.
  /// Throws error
  @override
  Future close(String? correlationId) async {
    if (_client != null) {
      await _client!.get_connection().close();
      _client = null;
    }
  }

  bool _checkOpened(String? correlationId) {
    if (!isOpen()) {
      throw InvalidStateException(
          correlationId, 'NOT_OPENED', 'Connection is not opened');
    }
    return true;
  }

  ///Makes a single attempt to acquire a lock by its key.
  ///It returns immediately a positive or negative result.
  ///
  /// - [correlationId]     (optional) transaction id to trace execution through call chain.
  /// - [key]               a unique lock key to acquire.
  /// - [ttl]               a lock timeout (time to live) in milliseconds.
  /// Return                Future that receives a lock result
  /// Throws error.
  @override
  Future<bool> tryAcquireLock(
      String? correlationId, String key, int ttl) async {
    if (!_checkOpened(correlationId)) return false;
    var result =
        await _client!.send_object(['SET', key, _lock, 'NX', 'PX', ttl]);
    return result == 'OK';
  }

  ///Releases prevously acquired lock by its key.
  ///
  /// - [correlationId]     (optional) transaction id to trace execution through call chain.
  /// - [key]               a unique lock key to release.
  /// Return          Future that receives an null for success.
  /// Throws error
  @override
  Future releaseLock(String? correlationId, String key) async {
    if (!_checkOpened(correlationId)) return;

    // Start transaction on key
    await _client!.send_object(['WATCH', key]);

    // Read and check if lock is the same
    var keyId = await _client!.send_object(['GET', key]);

    if (keyId == null) {
      await _client!.send_object(['UNWATCH']);
    }
    // Remove the lock if it matches
    if (keyId == _lock) {
      await _client!.send_object(['MULTI']);
      await _client!.send_object(['DEL', key]);
      await _client!.send_object(['EXEC']);
    } else {
      // Cancel transaction if it doesn't match
      await _client!.send_object(['UNWATCH']);
    }
  }
}


lib/src/cache/RedisCache.dart
import 'dart:async';

import 'package:redis/redis.dart' as redis;
import 'package:pip_services3_commons/pip_services3_commons.dart';
import 'package:pip_services3_components/pip_services3_components.dart';

///Distributed cache that stores values in Redis in-memory database.
///
///### Configuration parameters ###
///
/// - [connection(s)]:
///  - [discovery_key]:         (optional) a key to retrieve the connection from [IDiscovery](https://pub.dev/documentation/pip_services3_components/latest/pip_services3_components/IDiscovery-class.html)
///  - [host]:                  host name or IP address
///  - [port]:                  port number
///  - [uri]:                   resource URI or connection string with all parameters in it
/// - [credential(s)]:
///  - [store_key]:             key to retrieve parameters from credential store
///  - [username]:              user name (currently is not used)
///  - [password]:              user password
/// - [options]:
///  - [retries]:               number of retries (default: 3)
///  - [timeout]:               default caching timeout in milliseconds (default: 1 minute)
///  - [max_size]:              maximum number of values stored in this cache (default: 1000)
///
///### References ###
///
/// - *:discovery:*:*:1.0        (optional) [IDiscovery](https://pub.dev/documentation/pip_services3_components/latest/pip_services3_components/IDiscovery-class.html) services to resolve connection
/// - *:credential-store:*:*:1.0 (optional) Credential stores to resolve credential [ICredentialStore](https://pub.dev/documentation/pip_services3_components/latest/pip_services3_components/ICredentialStore-class.html)
///
///### Example ###
///```dart
///    var cache = RedisCache();
///    cache.configure(ConfigParams.fromTuples([
///      "host", "localhost",
///      "port", 6379
///    ]));
///
///   await cache.open("123");
///      ...
///
///    await cache.store("123", "key1", "ABC");
///    var value = await cache.retrieve("123", "key1");
///     // Result: "ABC"
///```
class RedisCache implements ICache, IConfigurable, IReferenceable, IOpenable {
  final _connectionResolver = ConnectionResolver();
  final _credentialResolver = CredentialResolver();

  int _timeout = 30000;
  int _retries = 3;

  redis.Command? _client;

  ///Creates a new instance of this cache.
  RedisCache();

  ///Configures component by passing configuration parameters.
  ///
  /// - [config]    configuration parameters to be set.
  @override
  void configure(ConfigParams config) {
    _connectionResolver.configure(config);
    _credentialResolver.configure(config);

    _timeout = config.getAsIntegerWithDefault('options.timeout', _timeout);
    _retries = config.getAsIntegerWithDefault('options.retries', _retries);
  }

  ///Sets references to dependent components.
  ///
  /// - [references] 	references to locate the component dependencies.
  @override
  void setReferences(IReferences references) {
    _connectionResolver.setReferences(references);
    _credentialResolver.setReferences(references);
  }

  ///Checks if the component is opened.
  ///
  ///Returns true if the component has been opened and false otherwise.
  @override
  bool isOpen() {
    return _client != null;
  }

  ///Opens the component.
  ///
  /// - [correlationId] 	(optional) transaction id to trace execution through call chain.
  /// Return 			Future that receives null no errors occured.
  /// Throws error
  @override
  Future open(String? correlationId) async {
    var connection = await _connectionResolver.resolve(correlationId);
    if (connection == null) {
      throw ConfigException(
          correlationId, 'NO_CONNECTION', 'Connection is not configured');
    }
    //credential = await _credentialResolver.lookup(correlationId);

    var redisConn = redis.RedisConnection();

    //TODO: Fix work with uri connection string and credentials
    // if (connection.getUri() != null) {
    //   var url = connection.getUri();
    // } else {
    var host = connection.getHost() ?? 'localhost';
    var port = connection.getPort() ?? 6379;
    _client = await redisConn.connect(host, port);

    // }

    // if (credential != null) {
    //   var password = credential.getPassword();
    // }
  }

  ///Closes component and frees used resources.
  ///
  /// - [correlationId] 	(optional) transaction id to trace execution through call chain.
  /// Return 			Future that receives null if no errors occured.
  /// Throws error
  @override
  Future close(String? correlationId) async {
    if (_client != null) {
      await _client!.get_connection().close();
      _client = null;
    }
  }

  bool _checkOpened(String? correlationId) {
    if (!isOpen()) {
      throw InvalidStateException(
          correlationId, 'NOT_OPENED', 'Connection is not opened');
    }
    return true;
  }

  ///Retrieves cached value from the cache using its key.
  ///If value is missing in the cache or expired it returns null.
  ///
  /// - [correlationId]     (optional) transaction id to trace execution through call chain.
  /// - [key]               a unique value key.
  /// Return                Future that receives cached value.
  /// Throws error
  @override
  Future<dynamic> retrieve(String? correlationId, String key) async {
    if (!_checkOpened(correlationId)) return;
    return await _client!.get(key);
  }

  ///Stores value in the cache with expiration time.
  ///
  /// - [correlationId]     (optional) transaction id to trace execution through call chain.
  /// - [key]               a unique value key.
  /// - [value]             a value to store.
  /// - [timeout]           expiration timeout in milliseconds.
  /// Return                Future that receives an 'OK' for success
  /// Throws error
  @override
  Future<dynamic> store(
      String? correlationId, String key, value, int timeout) async {
    if (!_checkOpened(correlationId)) return;
    return await _client!.send_object(['SET', key, value, 'PX', timeout]);
  }

  ///Removes a value from the cache by its key.
  ///
  /// - [correlationId]     (optional) transaction id to trace execution through call chain.
  /// - [key]               a unique value key.
  /// Return                Future function that receives an null for success
  /// Throws error
  @override
  Future<dynamic> remove(String? correlationId, String key) async {
    if (!_checkOpened(correlationId)) return;
    return await _client!.send_object(['DEL', key]);
  }
}


example/main.dart
import 'dart:io';
import 'package:pip_services3_commons/pip_services3_commons.dart';
import 'package:pip_services3_redis/pip_services3_redis.dart';

void main() async {
  final LOCK1 = 'lock_1';
  final LOCK2 = 'lock_2';

  RedisLock _lock;

  var host = Platform.environment['REDIS_SERVICE_HOST'] ?? 'localhost';
  var port = Platform.environment['REDIS_SERVICE_PORT'] ?? 6379;

  _lock = RedisLock();

  var config = ConfigParams.fromTuples(
      ['connection.host', host, 'connection.port', port]);
  _lock.configure(config);

  await _lock.open(null);

  // Try to acquire lock for the first time
  var result = await _lock.tryAcquireLock('123', LOCK1, 3000);
  print(result); // true

  // Try to acquire lock for the second time
  result = await _lock.tryAcquireLock('123', LOCK1, 3000); // false

  // Release the lock
  await _lock.releaseLock('123', LOCK1);

  // Acquire lock for the first time
  await _lock.acquireLock('123', LOCK2, 3000, 1000);
  // Acquire lock for the second time
  try {
    await _lock.acquireLock('123', LOCK2, 3000, 1000); // error
  } catch (err) {
    // error handle
  }

  // Release the lock
  await _lock.releaseLock('123', LOCK2);

  await _lock.close(null);
}
